sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java

```diff
@@ -508,40 +542,99 @@ private void captureSessionToken(RxDocumentServiceRequest request, Map<String, S
-                (this.defaultConsistencyLevel == ConsistencyLevel.SESSION &&
-                    // skip applying the session token when Eventual Consistency is explicitly requested
-                    // on request-level for data plane operations.
-                    // The session token is ignored on teh backend/gateway in this case anyway
-                    // and the session token can be rather large (even run in the 16 KB header length problem
-                    // on the gateway - so not worth sending when not needed
-                    (!request.isReadOnlyRequest() ||
-                        request.getResourceType() != ResourceType.Document ||
-                        !Strings.areEqual(requestConsistencyLevel, ConsistencyLevel.EVENTUAL.toString())));
+        boolean sessionConsistency = RequestHelper.getConsistencyLevelToUse(this.gatewayServiceConfigurationReader,
+            request) == ConsistencyLevel.SESSION;
 
         if (!Strings.isNullOrEmpty(request.getHeaders().get(HttpConstants.HttpHeaders.SESSION_TOKEN))) {
-            if (!sessionTokenApplicable || isMasterOperation(request.getResourceType(), request.getOperationType())) {
+            if (!sessionConsistency ||
+                (!request.isReadOnlyRequest() && request.getOperationType() != OperationType.Batch && !this.useMultipleWriteLocations)){
                 request.getHeaders().remove(HttpConstants.HttpHeaders.SESSION_TOKEN);
             }
-            return; //User is explicitly controlling the session.
+            return Mono.empty(); //User is explicitly controlling the session.
         }
 
-        if (!sessionTokenApplicable || isMasterOperation(request.getResourceType(), request.getOperationType())) {
-            return; // Only apply the session token in case of session consistency and when resource is not a master resource
+        if (!sessionConsistency ||
+            (!request.isReadOnlyRequest() && request.getOperationType() != OperationType.Batch && !this.useMultipleWriteLocations)) {
+            return Mono.empty();
+            // Only apply the session token in case of session consistency and if request is read only,
+            // apply token for write request only if batch operation or multi master
         }
 
-        //Apply the ambient session.
-        String sessionToken = this.sessionContainer.resolveGlobalSessionToken(request);
+        if (this.collectionCache != null && this.partitionKeyRangeCache != null) {
+            return this.collectionCache.resolveCollectionAsync(BridgeInternal.getMetaDataDiagnosticContext(request.requestContext.cosmosDiagnostics), request).
+                flatMap(collectionValueHolder -> {
 
-        if (!Strings.isNullOrEmpty(sessionToken)) {
-            headers.put(HttpConstants.HttpHeaders.SESSION_TOKEN, sessionToken);
+                    if(collectionValueHolder== null || collectionValueHolder.v == null) {
```

Comment:
for personal knowledge: when this could happen?

Reply:
If cache is stale or corrupted(for which we already handling), so I don't know the exact scenario , but added as @moderakh insisted https://github.com/Azure/azure-sdk-for-java/pull/25555#discussion_r767028606